### PR TITLE
3.0beta1 clusterfixes

### LINKED ISSUE #21130: [3.0 beta.1] Cluster Proxy Setup is broken
**Describe the bug**
The cluster proxy setup is dysfunctional in BBB 3.0 Beta.1, making it impossible to join a meeting. The breakage is caused by multiple issues in the setup documentation (https://docs.bigbluebutton.org/3.0/administration/cluster-proxy/), the HTML5 client code as well as nginx locations. In the following is a list of determined issues:

1. Doc: It is advised to change nginx locations in `/usr/share/bigbluebutton/nginx/bbb-html5.nginx`, but the mentioned locations are now found in `/usr/share/bigbluebutton/nginx/bbb-html5.nginx.static`. For consistency, the location for the guest lobby should also be added in the `.static`  file. Also, the example snippet of locations does not match the current contents of the file (e.g. no `@html5client` location anymore.
2. Doc: (might depend on a fix to HTML5 client code mentioned in 5) The `defaultHTML5ClientUrl=https://bbb-proxy.example.com/bbb-01/html5client` needs a trailing slash: `defaultHTML5ClientUrl=https://bbb-proxy.example.com/bbb-01/html5client/`
3. Doc: In the 3.0 doc, the following passage has been dropped: 
`Change the nginx $bbb_loadbalancer_node variable to the name of the load balancer node in /usr/share/bigbluebutton/nginx/loadbalancer.nginx to allow CORS requests:`
`set $bbb_loadbalancer_node https://bbb-proxy.example.com;`
However, the `loadbalancer.nginx` does still exist and in `presentation-slides.nginx` the variable `bbb_loadbalancer_node` variable is still referenced.
4. Doc: In the provided content for `/etc/bigbluebutton/bbb-graphql-middleware.yml` there is a superfluous ` ```yaml` which breaks the file if included.
5. HTML5 Client: When loading, a request is made to `https://bbb-proxy.example.com/bigbluebutton/api/`, without the hostname path (i.e. `https://meet.bbb.rlp.net/bbb-xyz/bigbluebutton/api`). This is caused by incorrect code in `bigbluebutton-html5/imports/api/bbb-web-api/index.ts`. A fix has been found by @schrd .
6. bbb-graphql-middleware/nginx: In `/usr/share/bigbluebutton/nginx/graphql.nginx` the `auth_request` breaks the CORS Pre-Flight.
7. bbb-web/nginx: In `/usr/share/bigbluebutton/nginx/web.nginx` the location for `/bigbluebutton/ping` is missing the CORS header, e.g. `add_header 'Access-Control-Allow-Origin' '*'`. It could also be added with `bbb_loadbalancer_node`, if the variable is set (see above).

**Version**
BigBlueButton Server 3.0.0-beta.1 (816)

### PR DESCRIPTION
**Pull request by schrd**
### What does this PR do?

Some things were still broken in cluster setups:

- generating the relative URL to the API endpoint
- some CORS headers were missing

### Closes Issue(s)

Closes #21130

### Motivation

Running multi node BBB without cluster setup feels wrong.

Thanks to @Ithanil and @danimo for investigating the issue

### CODE REVIEW

File: docs/docs/administration/cluster-proxy.md [hunk 1]
```diff
@@ -130,28 +130,23 @@ public:
...
-[Service]
-Environment=ROOT_URL=https://127.0.0.1/bbb-01/html5client
-Environment=DDP_DEFAULT_CONNECTION_URL=https://bbb-01.example.com/bbb-01/html5client
-```
-
-Prepend the mount point of bbb-html5 in all location sections except for the
-`location @html5client` section in `/usr/share/bigbluebutton/nginx/bbb-html5.nginx`:
+Copy `/usr/share/bigbluebutton/nginx/bbb-html5.nginx.static` to
+`/usr/share/bigbluebutton/nginx/bbb-html5-cluster.nginx` and prepend the mount
+point of bbb-html5 in all location sections:
```
Comment (Ithanil) [inline docs/docs/administration/cluster-proxy.md:135]: Just out of curiosity: How is the copied file included then? Given that `/usr/share/bigbluebutton/nginx/bbb-html5.nginx` explicitly includes `/usr/share/bigbluebutton/nginx/bbb-html5.nginx.static`, I would assume this file needs to be explicitly included as well.

**Comment by Ithanil** [comment]
Thanks for your PR, which is definitely an improvement upon the workarounds we had so far. I didn't test it yet, but would assume this PR indeed fixes points 1), 5), 6), 7) and probably 2) in my linked issue.

However, besides the comments I made in the code, would you mind commenting on points 3) and 4) in the issue? Even if 3) is indeed a non-issue, it would be good to have 4) fixed in this PR.

File: docs/docs/administration/cluster-proxy.md [hunk 2]
```diff
@@ -130,28 +130,23 @@ public:
...
-* `/etc/systemd/system/bbb-html5.service.d/cluster.conf`
-
-It should have the following content:
-
-```
-[Service]
-Environment=ROOT_URL=https://127.0.0.1/bbb-01/html5client
-Environment=DDP_DEFAULT_CONNECTION_URL=https://bbb-01.example.com/bbb-01/html5client
-```
-
```
[the remark this answers is not in the document]
Comment (schrd) [reply docs/docs/administration/cluster-proxy.md:144]: bbb-html5 is no longer a service. Meteor has been removed.

[the remark this answers is not in the document]
Comment (Ithanil) [reply docs/docs/administration/cluster-proxy.md:144]: I wasn't aware of the implications regarding bbb-html5 service, thanks.

**Comment by schrd** [comment]
> However, besides the comment I made in the code, would you mind commenting on points 3) and 4) in the issue? Even if 3) is indeed a non-issue, it would be good to have 4) fixed in this PR.

Number 3) is no longer required. As far as I can see the `$bbb_loadbalancer_node` variable is no longer used. A fix for 4) is provided in 124d527e4

**Comment by Ithanil** [comment]
> > However, besides the comment I made in the code, would you mind commenting on points 3) and 4) in the issue? Even if 3) is indeed a non-issue, it would be good to have 4) fixed in this PR.
> 
> Number 3) is no longer required. As far as I can see the `$bbb_loadbalancer_node` variable is no longer used. A fix for 4) is provided in [124d527](https://github.com/bigbluebutton/bigbluebutton/commit/124d527e49c2a37428bd6561feaac31f4823614e)

Thanks for addressing 4).

Regarding 3), the variable is still used in `/usr/share/bigbluebutton/nginx/presentation-slides.nginx`, as mentioned in the issue.

**Comment by schrd** [comment]
> Regarding 3), the variable is still used in `/usr/share/bigbluebutton/nginx/presentation-slides.nginx`, as mentioned in the issue.

Oh you're right. Sorry I missed that. I can imagine 2 solutions:

1. `add_header 'Access-Control-Allow-Origin' '*';`. There are currently no access checks implemented, that means that anyone who can access a `meeting_id` and `presentation_id` can download presentation files and rendered SVGs.
2. protect presentation assets by some nginx `auth_request` check. That probably involves `bbb-web` and needs access to session cookies.

@antobinary which one do you prefer?

**Comment by danimo** [comment]
In the spririt of moving forward, can we get this merged and deal with 8. and 9. of the original issue later?

**Comment by Ithanil** [comment]
> In the spririt of moving forward, can we get this merged and deal with 8. and 9. of the original issue later?

Good idea. But still, the last comment of @schrd regarding point 3 should be resolved.

**Comment by antobinary** [comment]
Hi @schrd 

> (...) I can imagine 2 solutions:
> add_header 'Access-Control-Allow-Origin' '*';. There are currently no access checks implemented, that means that anyone who can access a meeting_id and presentation_id can download presentation files and rendered SVGs.
> protect presentation assets by some nginx auth_request check. That probably involves bbb-web and needs access to session cookies.
> @antobinary which one do you prefer?

`add_header 'Access-Control-Allow-Origin' '*';` should be fine for now I think.
We have looked into limiting the visibility of the resources based on who has access to the session and (based on memory) it was not a super low hanging fruit.

---

Would you like us to merge this PR as-is for beta.2 (being wrapped up)?

**Comment by antobinary** [comment]
Merging as-is [so we can ship as part of BBB 3.0.0-beta.2] and we can continue in a new PR.

**Pull request merged**